### ISSUE TITLE
Command Chair support for HasCrew, HasPassengers and RecoverKerbal

### ISSUE THREAD

**Issue by jrossignol** [jrossignol/ContractConfigurator#355]
From @inigmatus on [the forums](http://forum.kerbalspaceprogram.com/threads/101604-1-0-5-Contract-Configurator-v1-8-1-2015-11-15?p=2311725&viewfull=1#post2311725):

> I have a boat with command seats for the pilot and passenger. I can load in the passenger in the command seat on launch of the craft (providing a cabin for them to spawn into and then have pilot and passenger manually board the command seats on the boat), or I can do this easier via the Take Command mod right on launch from SPH which allows me to spawn direct to the command seats. 

> Anyways, I run into two issues:

> 1. The passenger is not recognized as being in the command seat by HasPassengers
> 2. On recovery of the pilot (and subsequently the vessel he's piloting) the passenger remains. RecoverKerbal fails to recognize separate recoveries for pilot and passenger.

> Would be really nice to figure out a way to make boats and planes using just command seats, to complete tourism or passenger transport contracts for early game careers that only start with a command seat and no closed cockpits.

**Comment by jrossignol** [comment]
@ignimatus - I think I'll need some more details on this one.  Tried a test with HasCrew and RecoverKerbal - both seemed to work okay with a Kerbal in the jumpseat.

**Comment by inigmatus** [comment]
I'll post a bug.cfg and steps to duplicate when I get a moment.

**Comment by inigmatus** [comment]
Tested.

This is not a Command Seat issue at all.

1. Take Command has a bug whereby spawned Kerbals from contracts, appear invisible in their Command Seats. The same is true when simply transfering a Kerbal to a Command Seat via the transfer command, so that issue does not even appear to be contract related.

2. I want players the fun of flying Command Seat-only aircraft with paying passengers. But without using Take Command mod.  Two possibilties here I tested:

A: attach a cabin for the passengers to SpawnPassengers into. EVA them out and drop them into Command Seats. They are of course recognized by HasPassengers etc. No bug here.  This is what I think you tested and also concluded.

...but here is the scenario that fails:

B: rather than attaching a cabin, instead, create a SpawnVessel with Crew can EVA over to the waiting airplane (and put into a Command Seat or crew cabin, doesn't matter). HasPassengers does not recognize the Crew that was spawned via SpawnVessel as valid passengers. This is repeatable. Probably would require an enhancement.

For now, this prevents spawning vessels with crews that need rescue - hence why I currently use the work around to SpawnKerbal nearby the SpawnVessel, such as my rescue contract requires me to SpawnKerbal and then SpawnVessel a boat underneath them. 2 hours of testing. Sheesh. Man this stuff takes a lot of work ;). 

BugTest.cfg - test cfg
TouristBus.craft - SpawnVessel
Bug Test 2.craft - airplane with cabin and Command Seats

Download: http://www.mediafire.com/download/4v2jjbl5k7vnrf3/BugTest.zip

Simply modify the cfg as necessary, and test using the Bug Test 2 aircraft.  Crew spawned by SpawnVessel is not recognized by HasPassengers.

1. Load Bug Test contract
2. Load Jeb as pilot for the Bug Test 2 craft
3. Launch.
4. Switch to TouristBus and EVA Kerbals to Bug Test 2 aircraft cabin.
5. See HasKerbal status not recognizing passengers, and on Recovery, not being marked off either.

**Comment by inigmatus** [comment]
TLDR; SpawnVessel crew not recognized by HasPassengers.

**Comment by severedsolo** [comment]
Sounds to me like HasPassengers is working correctly... It only recognises passengers spawned through the passenger spawn behaviour. I think what you want is HasCrew

**Comment by inigmatus** [comment]
I'll test HasCrew sometime today.

**Comment by inigmatus** [comment]
Is there a way to reference by index in HasPassengers any crew spawned by SpawnVessel/Crew? 

I'm just trying to figure out a way to absolutely ensure one has picked up stranded crews?  Say for example I want to spawn STS-107 Kolumbia with a hurt wing, in orbit. Contract would require picking up the stranded crew and returning them to KSC. How could I check for this?  If I use HasCrew, then my current launch rescue crew would be counted in addition to the STS-107 crew, no matter how many Kerbals I decided to make as my launch rescue crew, thus defeating the HasCrew parameter intended to check for the STS-107 crew before I even pick them up. Without a means to reference by index Kerbals spawned by SpawnVessel/Crew, I don't currently see a way to account for them precisely. Would this be a possible enhancement request at this point?

I suppose the work around would be to specify their names precisely in SpawnVessel/Crew and then use HasCrew to reference their names as a check?  If so, how could I randomly generate names in SpawnVessel/Crew, and how could I reference them in a HasCrew parameter?

**Comment by jrossignol** [comment]
HasPassenger can index into SpawnVessel crew.  Just start counting from 0 at the first CREW node.

Probably need to move this around in the documentation, but if you want to use generated referenced in HasCrew instead, then look at https://github.com/jrossignol/ContractConfigurator/wiki/Global-Functions
```
DATA
{
    type = string
    kerbal1 = RandomKerbalName(Male)
    kerbal2 = RandomKerbalName(Female)
    kerbal3 = RandomKerbalName(Random())
}
```
I think the third one has the correct syntax, but not sure.  Then just reference them as:
```
PARAMETER
{
    type = HasCrew

    kerbal = @/kerbal1
    kerbal = @/kerbal2
    kerbal = @/kerbal3
}
```

**Comment by inigmatus** [comment]
Thanks! I'll try that and report back.

**Comment by inigmatus** [comment]
You said
"HasPassenger can index into SpawnVessel crew. Just start counting from 0 at the first CREW node."
Did you mean HasCrew can index into SpawnVessel crew? Not HasPassenger, right? I tried HasPassenger and it can't reference any index of any SpawnVessel crew members, but HasCrew doesn't have an index attribute.

I did your HasCrew example above and it works great to reference SpawnVessel/Crew kerbals by name assuming one defines their names ahead of time in a DATA node and use those variables in each Crew node.

http://imgur.com/a/lhXd7

Only thing is, and its prob minor, is that gender is not specified in SpawnVessel/Crew, so my random kerbal names generated in my DATA node get assigned to SpawnVessel/Crew kerbals of any gender. Any possible way to alleviate this? Perhaps this could be the focus of this Issue enhancement: add gender designations to SpawnVessel/Crew kerbals (and a support request asking how to design a DATA node to generate random names with a matching random gender to be used in the SpawnVessel/Crew creation behavior), or a way to reference by index SpawnVessel/Crew in HasKerbal.


... and all this to get as a work around for Take Command spawning Tourist and Applicant Kerbals invisible in a Command Seat.

**Comment by jrossignol** [comment]
SpawnVessel/CREW nodes now support gender.  It was developed before female Kerbals, and I forgot to go back and fix it when that feature came out (but I did fix SpawnKerbal).

**Comment by jrossignol** [comment]
This:
```
PARAMETER
{
    name = HasPassengers
    type = HasPassengers
    name = @/kerbal1
    name = @/kerbal2
    disableOnStateChange = false
}
```
Cannot possibly work - name is the name of the parameter, not of a kerbal.  However, this would work:
```
PARAMETER
{
    name = HasPassengers
    type = HasPassengers
    kerbal = @/kerbal1
    kerbal = @/kerbal2
    disableOnStateChange = false
}
```
And is undocumented.  So I'll fix the wiki for that.

**Comment by jrossignol** [comment]
Just going to be testing the index/count stuff now - the other option is the name stuff worked for you because it's the index/count that's working (with default values, it really *should* pick up the stuff from spawnvessel.

**Comment by jrossignol** [comment]
Just checked, and HasPassengers *can* index into the SpawnVessel/CREW...  but it doesn't make sense, as HasPassengers needs the Kerbals to be loaded on the launchpad.  So really, what needs to be done here is to allow HasCrew to use index/count (it currently doesn't).

**Comment by jrossignol** [comment]
Okay, I just realized, adding index/count is a really bad idea here, because it introduces a bad chicken/egg problem in that the Kerbals aren't actually created until the contract is accepted, which means we can't generate the contract parameter description until the player accepts the contract.  Messy.  So let's go with the random kerbal name generation stuff, much better.  And the best I can come up with for your question on how to keep gender/name consistent would be (untested, may have syntax errors):
```
DATA
{
    type = List<Gender>
    genders = [ Random(), Random(), Random() ]
}
DATA
{
    type = List<string>
    kerbalNames = [ RandomKerbalName(@genders.ElementAt(0), RandomKerbalName(@genders.ElementAt(1), RandomKerbalName(@genders.ElementAt(2) ]
}
...
    CREW
    {
        name = @/kerbalNames.ElementAt(0)
        gender = @/genders.ElementAt(0)
    }
...
```
You don't have to use arrays of course, but the idea is the same.

**Comment by jrossignol** [comment]
With that, I'm calling this big long issue closed.
